{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook was used to make Figure 1b, showing how modified data differs from canonical in raw space."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import bz2\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pysam\n",
    "\n",
    "path_fasta = '../tamatoa/data/lambda_phage_unix.fa'\n",
    "fasta = pysam.FastaFile(path_fasta)\n",
    "\n",
    "path_align_canon = './data/canon_eventalign.tsv.bz2'\n",
    "path_align_mod = './data/mod_eventalign.tsv.bz2'\n",
    "\n",
    "# path_align_canon = './data/canon_eventalign.tsv.bz2'\n",
    "# path_align_mod = './data/glu_pass_eventalign.tsv.bz2'\n",
    "# path_align_mod = './data/glu_fail_eventalign.tsv.bz2'\n",
    "# path_align_mod = './data/glu_AJG810_fail_c4241646_14_eventalign.tsv.bz2'\n",
    "\n",
    "# path_align_canon = './data/canon_eventalign.tsv.bz2'\n",
    "# path_align_mod = './data/gpc_pass_eventalign.tsv.bz2'\n",
    "# path_align_mod = './data/gpc_fail_eventalign.tsv.bz2'\n",
    "\n",
    "region_start, region_end = 12335, 12355\n",
    "region_start, region_end = 20600, 20640\n",
    "\n",
    "def load_event_align(path_align,reads=[]):\n",
    "    with bz2.open(path_align,mode='rb') as f_ea:\n",
    "\n",
    "        header = f_ea.readline().decode().split()\n",
    "        splitline = []\n",
    "        cur_read_id = 0\n",
    "        cur_read = []\n",
    "        last_pos = -1\n",
    "        for line in f_ea:\n",
    "            if len(reads) >= 20:\n",
    "                break\n",
    "            splitline = line.split()\n",
    "            if region_start <= int(splitline[1]) < region_end:\n",
    "                # print(splitline[3],splitline[1],splitline[-1])\n",
    "                # new read\n",
    "                # Ignore reverse strand results (ref kmer must match model kmer)\n",
    "                if splitline[2]!=splitline[9]:\n",
    "                    continue\n",
    "                if int(splitline[3]) != cur_read_id:\n",
    "                    # Don't add if didn't cover full target region\n",
    "                    if len(cur_read) == region_end-region_start:\n",
    "                        reads.append(cur_read)\n",
    "                    cur_read_id = int(splitline[3])\n",
    "                    cur_read = []\n",
    "                    # print(line)\n",
    "                if int(splitline[1]) != last_pos:\n",
    "                    cur_read.append([int(splitline[1])])\n",
    "                    last_pos = int(splitline[1])\n",
    "                cur_read[-1].extend(float(x) for x in splitline[-1].decode().split(','))\n",
    "\n",
    "    return reads\n",
    "\n",
    "reads_canon = load_event_align(path_align_canon,[])\n",
    "print('canon',len(reads_canon))\n",
    "reads_mod_CpG = load_event_align(path_align_mod,[])\n",
    "print('mods',len(reads_mod_CpG))\n",
    "\n",
    "# GpC data\n",
    "reads_mod_GpC = load_event_align('./data/gpc_pass_eventalign.tsv.bz2',[])\n",
    "reads_mod_GpC = load_event_align('./data/gpc_fail_eventalign.tsv.bz2',reads_mod_GpC)\n",
    "print('mods',len(reads_mod_GpC))\n",
    "\n",
    "\n",
    "### Gluc stuff, as that needed more data to plot properly\n",
    "#Load a lot of eventalign files due to little target coverage per batch in this sample\n",
    "reads_mod_GluC = load_event_align('./data/glu_fail_eventalign.tsv.bz2',[])\n",
    "for i in ['4','14','34','44','54','64','74','8','18','28','38','48','58','68','78']:\n",
    "    reads_mod_GluC = load_event_align('./data/glu_AJG810_fail_c4241646_'+i+'_eventalign.tsv.bz2',reads=reads_mod_GluC)\n",
    "    print(i,len(reads_mod_GluC))\n",
    "print('mods',len(reads_mod_GluC))\n",
    "\n",
    "# reads_mod = load_event_align('./data/glu_fail_eventalign.tsv.bz2')\n",
    "# for i in ['4','14']:\n",
    "#     reads_mod = load_event_align('./data/glu_AJG810_fail_c4241646_'+i+'_eventalign.tsv.bz2',reads=reads_mod)\n",
    "#     print(i,len(reads_mod))\n",
    "\n",
    "# reads_mod = []\n",
    "# for i in ['4','14','24','34','44','54','64','74','8','18','28','38','48','58','68','78']:\n",
    "#     reads_mod = load_event_align('./data/glu_AJG810_fail_c4241646_'+i+'_eventalign_mcpg.tsv.bz2',reads=reads_mod)\n",
    "#     print(i,len(reads_mod))\n",
    "# print('mods',len(reads_mod))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def plot_event_align(reads,my_color='k'):\n",
    "    # print(reads)\n",
    "    i = 0\n",
    "    for read in reads[:15]:\n",
    "        # Ignore reads that didn't cover the full area\n",
    "\n",
    "        # print(read)\n",
    "        x_values = []\n",
    "        y_values = []\n",
    "        for base in read:\n",
    "            x_values.extend(base[0] + np.arange(len(base)-1)/len(base)+1/len(base))\n",
    "            y_values.extend(np.array(base[1:]))#+i*50)\n",
    "        plt.plot(x_values,y_values,color=my_color,alpha=0.2,linewidth=.75)\n",
    "        i+=1\n",
    "\n",
    "# plot_type = 'glu'\n",
    "\n",
    "def plot_aligned_events(reads_mod,plot_type):\n",
    "    plt.figure(figsize=(8,4))\n",
    "    for pos in np.arange(region_start,region_end):\n",
    "        plt.axvline(pos,linestyle='--',linewidth=.75,color='k',alpha=.5)\n",
    "    plot_event_align(reads_canon,my_color='b')\n",
    "    # plot_event_align(reads_mod_CpG,my_color='r')\n",
    "    # plot_event_align(reads_mod_GpC,my_color='r')\n",
    "    # plot_event_align(reads_mod_GluC,my_color='r')\n",
    "    plot_event_align(reads_mod,my_color='r')\n",
    "    refseq = fasta.fetch('J02459.1')[region_start+2:region_end+2]\n",
    "    refseq_list = list(refseq)\n",
    "    if plot_type in ['cpg','glu']:\n",
    "        refseq_list[17] = 'C/M' # CpG\n",
    "    if plot_type == 'gpc':\n",
    "        refseq_list[14] = 'C/M' # GpC\n",
    "        refseq_list[19] = 'C/M' # GpC\n",
    "        refseq_list[24] = 'C/M' # GpC\n",
    "    plt.xticks(np.arange(region_start,region_end)+.5,labels=refseq_list)\n",
    "    plt.ylim(60,135)\n",
    "    # plt.xlim(region_start+12,region_end-15)\n",
    "    plt.xlim(region_start+11,region_end-16)\n",
    "    plt.xlabel('Reference sequence')\n",
    "    plt.ylabel('Measured current')\n",
    "    plt.plot([0,0],color='b',alpha=0.5,linewidth=2,label='Canonical')\n",
    "    label = 'Modified'\n",
    "    if plot_type == 'cpg': label='CpG methylated'\n",
    "    if plot_type == 'gpc': label='GpC methylated'\n",
    "    if plot_type == 'glu': label='CpG glucosylated'\n",
    "\n",
    "    plt.plot([0,0],color='r',alpha=0.5,linewidth=2,label=label)\n",
    "    plt.title('Comparison between canonical and '+label+' data')\n",
    "    plt.legend(loc='upper right')\n",
    "    plt.tight_layout()\n",
    "    plt.savefig('./data/event_align_'+plot_type+'.pdf') #CpG\n",
    "\n",
    "# plt.savefig('./data/event_align_glu_mcpg.pdf') # GluC with methylated ref\n",
    "# plt.savefig('./data/event_align_all.pdf') \n",
    "# print('J02459.1',region_start+2,region_end+2,refseq)\n",
    "    \n",
    "plot_aligned_events(reads_mod_CpG,'cpg')\n",
    "plot_aligned_events(reads_mod_GpC,'gpc')\n",
    "plot_aligned_events(reads_mod_GluC,'glu')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "plt.figure(figsize=(8,4))\n",
    "# plot_event_align(reads_canon,my_color='b')\n",
    "print((reads_canon[0]))\n",
    "plt.plot(np.concatenate([x[1:] for x in reads_canon[0]]))\n",
    "refseq = fasta.fetch('J02459.1')[region_start+2:region_end+2]\n",
    "plt.xlabel('Reference sequence')\n",
    "plt.ylabel('Measured current')\n",
    "plt.title('Raw read data aligned to reference sequence')\n",
    "plt.legend()\n",
    "plt.tight_layout()\n",
    "plt.savefig('./data/raw_data.pdf')\n",
    "print('J02459.1',region_start+2,region_end+2,refseq)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "nanopore",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
